rpc/lib/types/types.go

```diff
@@ -13,17 +13,63 @@ import (
 	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
 )
 
+// a wrapper to emulate a sum type: jsonrpcid = string | int
+// TODO: refactor when Go 2.0 arrives https://github.com/golang/go/issues/19412
+type jsonrpcid interface {
+	isJSONRPCID()
+}
+
+// JSONRPCStringID a wrapper for JSON-RPC string IDs
+type JSONRPCStringID string
+
+func (JSONRPCStringID) isJSONRPCID() {}
+
+// JSONRPCIntID a wrapper for JSON-RPC integer IDs
+type JSONRPCIntID int
+
+func (JSONRPCIntID) isJSONRPCID() {}
+
 //----------------------------------------
 // REQUEST
 
 type RPCRequest struct {
 	JSONRPC string          `json:"jsonrpc"`
-	ID      string          `json:"id"`
+	ID      jsonrpcid       `json:"id"`
 	Method  string          `json:"method"`
 	Params  json.RawMessage `json:"params"` // must be map[string]interface{} or []interface{}
 }
 
-func NewRPCRequest(id string, method string, params json.RawMessage) RPCRequest {
+// UnmarshalJSON custom JSON unmarshalling due to jsonrpcid being string or int
+func (request *RPCRequest) UnmarshalJSON(data []byte) error {
+	type AliasStr RPCRequest
```

Comment:
We should Unmarshal interface{} and then use switch for different possible types

```go
switch ID.(type) {
  case string:
           XXX
  case int:
           YYY
  default:
    return fmt.Errorf("Unknown ID type. Got %T, expected string or int", ID)
```

Reply:
@melekes https://github.com/tendermint/tendermint/pull/2811#discussion_r232934518 ?

Reply:
yes! also you can do without intermediate struct

```go
unsafeResp := &struct {
            JSONRPC string          `json:"jsonrpc"`
            ID      interface{}     `json:"id"`
            Method  string          `json:"method"`
            Params  json.RawMessage `json:"params"` // must be map[string]interface{} or []interface{}
        }{}
```

you just use the original one

Reply:
@melekes @ebuchman I've changed it here: https://github.com/tendermint/tendermint/pull/2811/commits/f2a71d7b9dd636816c0da35384779390ff1a3338#diff-f595cd0be03b00ec17ec43d65afb7780

I'm not that experienced with Go, so not sure how to do it without the intermediate struct -- or did you mean that `jsonrpcid` will be an empty interface and it should get rid of the extra ID type wrappers (`JSONRPCStringID`/`JSONRPCIntID`)?